### PR TITLE
routing subsystem refactoring

### PR DESCRIPTION
**Pull request by fafhrd91**
routing subsystem refactoring

1. Route object is responsible for path only (endpoint)
2. handler moved to View object. we can add more matching checks to View, like specific headers, cookies, etc. right now we match only on method.
3. added system routes that match HTTPException, so we can register view for specific system route, i.e. HTTPNotFound, or HTTPForbidden, etc.

this is first version, for discussion
it is not backward compatible, we can make it backward compatible later when we decide if we want to use this changes.

@asvetlov @kxepal  thoughts, ides, etc

### CODE REVIEW

File: aiohttp/hdrs.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 METH_POST = upstr('POST')
 METH_PUT = upstr('PUT')
 METH_TRACE = upstr('TRACE')
+METH_PURGE = upstr('PURGE')
```
Comment (kxepal) [inline aiohttp/hdrs.py:14]: Looks like unrelated change. PURGE is used for Varnish needs, but what's it purpose for aiohttp?

File: aiohttp/web_urldispatcher.py [hunk 2]
```diff
@@ -91,53 +84,113 @@ def _append_query(url, query):
             return url
 
 
-class PlainRoute(Route):
+class ViewableRoute(Route):
+
+    def __init__(self, name, *, expect_handler=None):
+        super().__init__(name, expect_handler=expect_handler)
+        self._views = []
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:91]: I believe that list is temporary solution as on practice here should be tree-like thing because Views may serve various content type depending on Accept-\* headers.

File: aiohttp/web_urldispatcher.py [hunk 3]
```diff
@@ -91,53 +84,113 @@ def _append_query(url, query):
...
+
+    def match_view(self, request, view_name, match_dict):
+        method = request.method
+        allowed_methods = set()
+
+        for view in self._views:
+            if (view.name == view_name and
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:98]: Another issue with the list choice is the silent collision here. Assume the following (pseudo)-code:

```
route = app.add_route('/foo/bar')
route.add_view(METH_ANY, handle_any)
route.add_view(METH_GET, handle_get)
```

The idea behind is to catch GET requests with the special hander, but everything else with some generic one. But here it wouldn't work because METH_ANY would have higher resolution order (which is depended on declaration one) while technically, it should be used as fallback case when no direct method match is possible.

File: aiohttp/web_urldispatcher.py [hunk 4]
```diff
@@ -91,53 +84,113 @@ def _append_query(url, query):
...
+
+    def url(self, **kwargs):
+        raise RuntimeError(".url() is not allowed for SystemRoute")
+
+    def match(self, request):
+        pass
+
+    def handler(self, request, exc):
+        request.match_info['exception'] = exc
+        if self._view or not None:
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: Did you mean `if self_view is not None`?

File: aiohttp/web_urldispatcher.py [hunk 5]
```diff
@@ -157,10 +209,16 @@ def __init__(self, name, prefix, directory, *,
...
                 "No directory exists at '{}'".format(self._directory))
 
-    def match(self, path):
+    def match(self, request):
+        path = request.raw_path
         if not path.startswith(self._prefix):
             return None
-        return {'filename': path[self._prefix_len:]}
+
+        if request.method != hdrs.METH_GET:
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:217]: Unrelated note, but HEAD method support here would be nice.

File: aiohttp/web_urldispatcher.py [hunk 6]
```diff
@@ -169,7 +227,7 @@ def url(self, *, filename, query=None):
         return self._append_query(url, query)
 
     @asyncio.coroutine
-    def handle(self, request):
+    def handler(self, request):
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:230]: I was though that StaticRoute would be a good builtin example how to use views. Or you didn't find this idea good as well?

File: aiohttp/web_urldispatcher.py [hunk 7]
```diff
@@ -213,35 +271,32 @@ def handle(self, request):
...
-        super().__init__(hdrs.METH_ANY, None, None)
-        self._status = status
-        self._reason = reason
-
-    def url(self, **kwargs):
-        raise RuntimeError(".url() is not allowed for SystemRoute")
+    def __init__(self, handler, name='', method=hdrs.METH_ANY):
+        self._name = name
+        self._method = upstr(method)
+        self._handler = handler
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:288]: Shouldn't here be that code now assuming that only View now is cares about handler type and coroutine nature:

```
         assert callable(handler), handler
         if (not asyncio.iscoroutinefunction(handler) and
                 not inspect.isgeneratorfunction(handler)):
             handler = asyncio.coroutine(handler)
```

File: aiohttp/web_urldispatcher.py [hunk 8]
```diff
@@ -353,24 +401,29 @@ def register_route(self, route):
...
             handler = asyncio.coroutine(handler)
 
-        method = upstr(method)
-        if method not in self.METHODS:
-            raise ValueError("{} is not allowed HTTP method".format(method))
+        if route is None:
+            raise ValueError('Route name is required')
+
+        if route not in self._routes:
+            raise ValueError('Route is not found {}', route)
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: Missed format?

File: aiohttp/web_urldispatcher.py (hunk 2 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py:91]: i expect that be O(n) match for each view.

File: aiohttp/web_urldispatcher.py (hunk 3 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py:98]: i do not think this is a problem in general. later can add some warning or reordering for view, whatever.

File: aiohttp/web_urldispatcher.py (hunk 4 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py]: you are right

File: aiohttp/web_urldispatcher.py [hunk 9]
```diff
@@ -353,24 +401,29 @@ def register_route(self, route):
                 self._routes[name] = route
         self._urls.append(route)
 
-    def add_route(self, method, path, handler,
-                  *, name=None, expect_handler=None):
-
-        if not path.startswith('/'):
-            raise ValueError("path should be started with /")
+    def add_view(self, method, handler, *, name='', route=None):
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:409]: Hm, I was thought that `route` argument should take Route instance, not the name before read the tests. Does it make same to handle both here?

File: aiohttp/web_urldispatcher.py (hunk 6 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py:230]: static route is so specific, there is no reason to make it more generic

File: aiohttp/web_urldispatcher.py (hunk 7 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py:288]: yes, we should check handler in View

File: aiohttp/web_urldispatcher.py (hunk 9 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py:409]: i thought about opposite, if you have route object, you can just call `Route.add_view()`

File: aiohttp/web_urldispatcher.py (hunk 8 above)
Comment (fafhrd91) [inline aiohttp/web_urldispatcher.py]: `route` is string, maybe it is better to rename to `route_name`

Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: Good idea. Also will solve str/Route confusion.

Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: But actually I mean:

```
raise ValueError('Route is not found {}'.format(route))
```

Or did I miss some new Python feature?

File: aiohttp/web_urldispatcher.py (hunk 9 above)
Comment (kxepal) [inline aiohttp/web_urldispatcher.py:409]: Indeed, you're right.

**Comment by kxepal** [comment]
To not write "ok"  for all inline comments, will say that here (: 
Looks good for me, at least after reading. Will try to play tomorrow with the fresh head, but I think that's a good step forward with the large room for the improvements.

Also, I didn't get how to register system routes (tests also missed example of such case), but may be I should take a look a bit closer.

**Comment by fafhrd91** [comment]
System routes are not clear yet. At least Should we allow to replace existing system routes or just allow to set view on it.

**Comment by kxepal** [comment]
@fafhrd91 hm...I think that exceptions are the same base for views as methods:

```
route = router.add_route('foo', '/bar/baz')
route.add_view(METH_GET, handle_get)
route.add_view(HTTPNotFound, handle_404)
route.add_view(HTTPBadRequest, handle_400)
...
```

While there are also "default" system routes (actually views) to catch and process these errors when user route doesn't provides own handlers or accidentally reraise the error from own handler.

**Comment by fafhrd91** [comment]
Oh! I really like it.

**Comment by asvetlov** [comment]
I like idea very much but please don't rush with merging.
Let me review carefully.

**Comment by fafhrd91** [comment]
sure, it is too early for merging.

some ideas.
we can method default to METH_ANY and make it optional.
also we can extract path leaf and use it as view name, something like `/path-to-comething/index.html` becomes route with path `/path-to-comething/` and view with name `index.html`

would be cool to make this to work `route.add_view(HTTPBadRequest, handle_400)`, we may need context for view registration, but context adds so many complications.

**Comment by asvetlov** [comment]
I'm also thinking about nested applications and routers.

**Comment by tark-hidden** [comment]
Would you think about routing system based on route name? I mean...

```
app.add_route('foo', '/bar/baz')
app.add_view('foo', METH_GET, handler)
app.add_view('foo', METH_POST, handler)
```

...where `foo` is route_name. Yes, Pyramid-like, but... Pros: less variables, simpler code, easy to write. Cons: None.

**Comment by kxepal** [comment]
@tark-hidden This PR allows that, though you'll have to specify route name explicitly via keyword argument:

```
app.add_route('foo', '/bar/baz')
app.add_view(METH_GET, handler, route='foo')
app.add_view(METH_POST, handler, route='foo')
```

However, if route name is really mandatory for `add_view` method, indeed, why not to pass it with the first argument?

**Comment by tark-hidden** [comment]
Ah, sorry! I've read an example code with `route = router.add_route('foo', '/bar/baz')` and `route.add_view(...)`. If I can use `route` variable as route name without unnecessary variables, it is fine! It that case placement of this parameter doesn't matter, and `method` as first parameter is better choice.

**Comment by fafhrd91** [comment]
@asvetlov did you have a chance to review this PR

**Comment by asvetlov** [comment]
Sorry, I still have no time for exhausting review. I'm busy with preparing to europython. Maybe next week?

**Comment by asvetlov** [comment]
As I see new route system is not compatible with existing one, right?

So maybe better to add new router class (`UrlDispather2`/`web_urldispatcher2.py`)?

I support splitting into route/view BTW.

**Comment by auvipy** [comment]
@fafhrd91 This branch has conflicts that must be resolved

**Comment by fafhrd91** [comment]
I do not have time for this change. Anyone wants to work on this change?

**Comment by auvipy** [comment]
count me on for this then

**Comment by fafhrd91** [comment]
Please coordinate this effort with @asvetlov

**Comment by auvipy** [comment]
hi @asvetlov

**Comment by asvetlov** [comment]
Hello.
The PR is not backward compatible with existing code.
There are two options:
1. Make it compatible (preferable solution)
2. Move new solution into new `UrlDispatcher2` class with deprecating `UrlDispatcher` in future aiohttp releases (adds a burden of supporting two routers in parallel for a while, about 9-12 months long, which is really maintenance nightmare).

**Comment by fafhrd91** [comment]
i think we should maintain compatibility.

**Comment by auvipy** [comment]
i'm up for point 1

**Comment by asvetlov** [comment]
@auvipy please let me know when you will have something to discuss.

**Comment by auvipy** [comment]
@asvetlov at gitter? or mail or here?

**Comment by asvetlov** [comment]
I think github is better tool.
We will left a discussion history at least.

**Comment by auvipy** [comment]
ok let me know what u think should be done? I am up for adding a django like deprication utlis module for better management of feature deprication and related stuffs @ aiohttp an then add/depreicate features according to roadmape design plans.

**Comment by asvetlov** [comment]
Try to figure out first what changes are required for adding new functionality without breaking any existing test.

**Comment by auvipy** [comment]
OK

**Comment by asvetlov** [comment]
Done by #727
